{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dcbae58e-ac49-4541-beea-995b266f76f1",
   "metadata": {},
   "source": [
    "# Tratamiento de datos con Pandas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b6480b2-d958-425d-a96f-3f2021a9d6eb",
   "metadata": {},
   "source": [
    "Pandas es una biblioteca de Python que agrega muchas funciones para manipular datos en forma tabular. Permite filtrar datos según los valores de una o más columnas, ordenas los datos, crear nuevas variables, etc."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62263358-5ca0-4417-8220-a18445a8538c",
   "metadata": {},
   "source": [
    "## Cargar las libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7517e5b-ed30-4a98-bfeb-22f970bda20f",
   "metadata": {},
   "source": [
    "Si al intentar cargas Pandas se genera un error que dice \"No module named 'pandas'\", es que no está instalada la library. Esto se puede hacer desde una temrinal del sistema operativo, también se puede abrir una terminal en JupyterLab (File > New > Terminal) y desde ahi ejecutar:\n",
    "conda install conda-forge::pandas\n",
    "o\n",
    "pip install pandas\n",
    "\n",
    "Si hicieron la instalación siguiendo las instrucciones del curso, el primer comando va a funcionar bien. Para instalaciones que no dependan de conda o minoconda, usar \"pip\". Si instalaron conda, en lugar d eminicnda, ya deberían tener Pandas instalado.\n",
    "\n",
    "También es probable que tengan instalar openpyxl:\n",
    "conda install conda-forge::openpyxl\n",
    "o\n",
    "pip install openpyxl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "38bfeb0c-2b1e-4c9d-85a1-4450b33e131a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a91bb8fa-cb2b-46b8-8f66-ff9e67d10266",
   "metadata": {},
   "source": [
    "### Lectura de archivos Excel\n",
    "Pandas incluye funciones para importar archivos en formato Excel. Es importante que en el archivo Excel los datos estén orgnizados en formato de tabla (ver los dos archivos de ejemplo)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f422396d-151e-46b6-bcf1-17d10eb1e672",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_ncbi = pd.read_excel(\"ncbi_dataset.xlsx\")\n",
    "genes_crec = pd.read_excel(\"growthScores.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "783a9662-e807-4915-be9e-7c8363b0d5f8",
   "metadata": {},
   "source": [
    "Los archivos Excel se leen en unas estructuras de datos propias de Pandas llamadas DataFrames. Estas estructuras son equivalentes a los dataframes de R, y Pandas implementa funcionalidad similares, aunque no idénticas, a las del paquete dplyr de R."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "01060b3e-7589-4c9b-8560-c034056ee54f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pandas.core.frame.DataFrame"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(genes_ncbi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b62c22f4-459a-4573-b6ae-8f135c19ee32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        Accession  Begin    End Chromosome Orientation  \\\n",
      "0     NC_001133.9   1807   2169          I       minus   \n",
      "1     NC_001133.9   2480   2707          I        plus   \n",
      "2     NC_001133.9   7235   9016          I       minus   \n",
      "3     NC_001133.9  11565  11951          I       minus   \n",
      "4     NC_001133.9  12046  12426          I        plus   \n",
      "...           ...    ...    ...        ...         ...   \n",
      "6472  NC_001224.1  78089  78162         MT       minus   \n",
      "6473  NC_001224.1  78533  78608         MT        plus   \n",
      "6474  NC_001224.1  79213  80022         MT        plus   \n",
      "6475  NC_001224.1  85035  85112         MT        plus   \n",
      "6476  NC_001224.1  85295  85777         MT        plus   \n",
      "\n",
      "                                Name Symbol    Gene_ID       Gene_Type  \\\n",
      "0                  seripauperin PAU8   PAU8   851229.0  protein-coding   \n",
      "1            uncharacterized protein    NaN  1466426.0  protein-coding   \n",
      "2             putative permease SEO1   SEO1   851230.0  protein-coding   \n",
      "3            uncharacterized protein    NaN   851232.0  protein-coding   \n",
      "4            uncharacterized protein    NaN   851233.0  protein-coding   \n",
      "...                              ...    ...        ...             ...   \n",
      "6472                        tRNA-Thr    NaN   854625.0            tRNA   \n",
      "6473                        tRNA-Val    NaN   854626.0            tRNA   \n",
      "6474  cytochrome c oxidase subunit 3   COX3   854627.0  protein-coding   \n",
      "6475                        tRNA-Met    NaN   854628.0            tRNA   \n",
      "6476                            RPM1   RPM1  9164989.0           ncRNA   \n",
      "\n",
      "     Transcripts_accession Protein_accession  Protein_length  Locus_tag  \n",
      "0           NM_001180043.1       NP_009332.1           120.0    YAL068C  \n",
      "1           NM_001184582.1       NP_878038.1            75.0  YAL067W-A  \n",
      "2           NM_001178208.1       NP_009333.1           593.0    YAL067C  \n",
      "3           NM_001179897.1       NP_009335.1           128.0    YAL065C  \n",
      "4           NM_001180042.1       NP_009336.1           126.0  YAL064W-B  \n",
      "...                    ...               ...             ...        ...  \n",
      "6472                   NaN               NaN             NaN  tT(UAG)Q2  \n",
      "6473                   NaN               NaN             NaN   tV(UAC)Q  \n",
      "6474                   NaN       NP_009328.1           269.0      Q0275  \n",
      "6475                   NaN               NaN             NaN  tM(CAU)Q2  \n",
      "6476                   NaN               NaN             NaN      Q0285  \n",
      "\n",
      "[6477 rows x 13 columns]\n"
     ]
    }
   ],
   "source": [
    "print(genes_ncbi)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a468a552-97f0-4c47-993e-fb4a970c90fb",
   "metadata": {},
   "source": [
    "## Seleccionar columnas y filas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4698fb63-b620-4b77-b1b0-4519675811e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Accession', 'Begin', 'End', 'Chromosome', 'Orientation', 'Name',\n",
       "       'Symbol', 'Gene_ID', 'Gene_Type', 'Transcripts_accession',\n",
       "       'Protein_accession', 'Protein_length', 'Locus_tag'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_ncbi.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "336883e1-d4db-47ef-acda-4b15816bfa78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       NC_001133.9\n",
       "1       NC_001133.9\n",
       "2       NC_001133.9\n",
       "3       NC_001133.9\n",
       "4       NC_001133.9\n",
       "           ...     \n",
       "6472    NC_001224.1\n",
       "6473    NC_001224.1\n",
       "6474    NC_001224.1\n",
       "6475    NC_001224.1\n",
       "6476    NC_001224.1\n",
       "Name: Accession, Length: 6477, dtype: object"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_ncbi['Accession']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b776e78-bc4d-4ecd-a74c-deb6e3c494a1",
   "metadata": {},
   "source": [
    "Prestar atención a cómo se seleccionan dos o más columnas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "eaf56f33-e2d9-4980-bb6c-6fc38a0c5ba1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6472</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6473</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6474</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6475</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6476</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6477 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Accession Chromosome\n",
       "0     NC_001133.9          I\n",
       "1     NC_001133.9          I\n",
       "2     NC_001133.9          I\n",
       "3     NC_001133.9          I\n",
       "4     NC_001133.9          I\n",
       "...           ...        ...\n",
       "6472  NC_001224.1         MT\n",
       "6473  NC_001224.1         MT\n",
       "6474  NC_001224.1         MT\n",
       "6475  NC_001224.1         MT\n",
       "6476  NC_001224.1         MT\n",
       "\n",
       "[6477 rows x 2 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_ncbi[['Accession', 'Chromosome']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce9965ca-41f5-4242-9255-93f0d616122f",
   "metadata": {},
   "source": [
    "También se pueden referir las columnas por número. En este caso hay que usar el método, *iloc*, y luego indicar rangos de filas y de columnas, separados por comas. En el ejemplo de más abajo \"[:,0:3]\" quiere decir, queremos ver todas las filas (\":\"), luego queremos ver las columnas 0, 1 y 2. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "317df718-9eb3-484c-8287-7b09f3cc993e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Begin</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>1807</td>\n",
       "      <td>2169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>2480</td>\n",
       "      <td>2707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>7235</td>\n",
       "      <td>9016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>11565</td>\n",
       "      <td>11951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>12046</td>\n",
       "      <td>12426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6472</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>78089</td>\n",
       "      <td>78162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6473</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>78533</td>\n",
       "      <td>78608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6474</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>79213</td>\n",
       "      <td>80022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6475</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>85035</td>\n",
       "      <td>85112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6476</th>\n",
       "      <td>NC_001224.1</td>\n",
       "      <td>85295</td>\n",
       "      <td>85777</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6477 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Accession  Begin    End\n",
       "0     NC_001133.9   1807   2169\n",
       "1     NC_001133.9   2480   2707\n",
       "2     NC_001133.9   7235   9016\n",
       "3     NC_001133.9  11565  11951\n",
       "4     NC_001133.9  12046  12426\n",
       "...           ...    ...    ...\n",
       "6472  NC_001224.1  78089  78162\n",
       "6473  NC_001224.1  78533  78608\n",
       "6474  NC_001224.1  79213  80022\n",
       "6475  NC_001224.1  85035  85112\n",
       "6476  NC_001224.1  85295  85777\n",
       "\n",
       "[6477 rows x 3 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_ncbi.iloc[:,0:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10ceb522-dde6-4941-9f38-96a510761294",
   "metadata": {},
   "source": [
    "En el ejemplo que sigue indicamos las primeras 5 filas y luego las columnas 0, 2, 4, y 6:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "1e04ba50-75b8-4d6a-b9c7-33af9d2890f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>End</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>2169</td>\n",
       "      <td>minus</td>\n",
       "      <td>PAU8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>2707</td>\n",
       "      <td>plus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>9016</td>\n",
       "      <td>minus</td>\n",
       "      <td>SEO1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>11951</td>\n",
       "      <td>minus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_001133.9</td>\n",
       "      <td>12426</td>\n",
       "      <td>plus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Accession    End Orientation Symbol\n",
       "0  NC_001133.9   2169       minus   PAU8\n",
       "1  NC_001133.9   2707        plus    NaN\n",
       "2  NC_001133.9   9016       minus   SEO1\n",
       "3  NC_001133.9  11951       minus    NaN\n",
       "4  NC_001133.9  12426        plus    NaN"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_ncbi.iloc[:5,[0, 2, 4, 6]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6df7b803-45e4-49db-bc16-2a3e748f209b",
   "metadata": {},
   "source": [
    "## Filtrado de datos"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0416776-050e-45ba-a4f7-5f1d29183792",
   "metadata": {},
   "source": [
    "Con el método *loc* se pueden filtrar datos. Por ejemplo, para recuperar los datos sólo del cromosoma V:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "11425e51-f319-4341-a842-070898b0d35f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Begin</th>\n",
       "      <th>End</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Name</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Gene_Type</th>\n",
       "      <th>Transcripts_accession</th>\n",
       "      <th>Protein_accession</th>\n",
       "      <th>Protein_length</th>\n",
       "      <th>Locus_tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1522</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>264</td>\n",
       "      <td>4097</td>\n",
       "      <td>V</td>\n",
       "      <td>minus</td>\n",
       "      <td>Y' element ATP-dependent helicase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856630.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001184302.1</td>\n",
       "      <td>NP_010835.1</td>\n",
       "      <td>1277.0</td>\n",
       "      <td>YEL077C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1523</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>4185</td>\n",
       "      <td>5114</td>\n",
       "      <td>V</td>\n",
       "      <td>minus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856632.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001180855.1</td>\n",
       "      <td>NP_010836.1</td>\n",
       "      <td>216.0</td>\n",
       "      <td>YEL076C-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1524</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>4464</td>\n",
       "      <td>5114</td>\n",
       "      <td>V</td>\n",
       "      <td>minus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856633.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001178891.1</td>\n",
       "      <td>NP_010837.1</td>\n",
       "      <td>216.0</td>\n",
       "      <td>YEL076C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>5345</td>\n",
       "      <td>5713</td>\n",
       "      <td>V</td>\n",
       "      <td>minus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856634.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001178890.1</td>\n",
       "      <td>NP_010839.1</td>\n",
       "      <td>122.0</td>\n",
       "      <td>YEL075C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>7230</td>\n",
       "      <td>7553</td>\n",
       "      <td>V</td>\n",
       "      <td>minus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856636.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001178888.1</td>\n",
       "      <td>NP_010841.1</td>\n",
       "      <td>107.0</td>\n",
       "      <td>YEL073C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1835</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>568040</td>\n",
       "      <td>568759</td>\n",
       "      <td>V</td>\n",
       "      <td>plus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856938.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001348829.1</td>\n",
       "      <td>NP_001335771.1</td>\n",
       "      <td>239.0</td>\n",
       "      <td>YER188W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1836</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>569608</td>\n",
       "      <td>569907</td>\n",
       "      <td>V</td>\n",
       "      <td>minus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1466538.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001184633.1</td>\n",
       "      <td>NP_878070.1</td>\n",
       "      <td>99.0</td>\n",
       "      <td>YER188C-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1837</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>571155</td>\n",
       "      <td>571523</td>\n",
       "      <td>V</td>\n",
       "      <td>plus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856939.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001179079.3</td>\n",
       "      <td>NP_011116.3</td>\n",
       "      <td>122.0</td>\n",
       "      <td>YER189W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1838</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>571480</td>\n",
       "      <td>576525</td>\n",
       "      <td>V</td>\n",
       "      <td>plus</td>\n",
       "      <td>Y' element ATP-dependent helicase protein 1 co...</td>\n",
       "      <td>YRF1-2</td>\n",
       "      <td>856940.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001179080.3</td>\n",
       "      <td>NP_011117.3</td>\n",
       "      <td>1681.0</td>\n",
       "      <td>YER190W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1839</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>575680</td>\n",
       "      <td>576162</td>\n",
       "      <td>V</td>\n",
       "      <td>minus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4594642.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001184590.1</td>\n",
       "      <td>NP_001073292.1</td>\n",
       "      <td>160.0</td>\n",
       "      <td>YER190C-B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>318 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Accession   Begin     End Chromosome Orientation  \\\n",
       "1522  NC_001137.3     264    4097          V       minus   \n",
       "1523  NC_001137.3    4185    5114          V       minus   \n",
       "1524  NC_001137.3    4464    5114          V       minus   \n",
       "1525  NC_001137.3    5345    5713          V       minus   \n",
       "1526  NC_001137.3    7230    7553          V       minus   \n",
       "...           ...     ...     ...        ...         ...   \n",
       "1835  NC_001137.3  568040  568759          V        plus   \n",
       "1836  NC_001137.3  569608  569907          V       minus   \n",
       "1837  NC_001137.3  571155  571523          V        plus   \n",
       "1838  NC_001137.3  571480  576525          V        plus   \n",
       "1839  NC_001137.3  575680  576162          V       minus   \n",
       "\n",
       "                                                   Name  Symbol    Gene_ID  \\\n",
       "1522                  Y' element ATP-dependent helicase     NaN   856630.0   \n",
       "1523                            uncharacterized protein     NaN   856632.0   \n",
       "1524                            uncharacterized protein     NaN   856633.0   \n",
       "1525                            uncharacterized protein     NaN   856634.0   \n",
       "1526                            uncharacterized protein     NaN   856636.0   \n",
       "...                                                 ...     ...        ...   \n",
       "1835                            uncharacterized protein     NaN   856938.0   \n",
       "1836                            uncharacterized protein     NaN  1466538.0   \n",
       "1837                            uncharacterized protein     NaN   856939.0   \n",
       "1838  Y' element ATP-dependent helicase protein 1 co...  YRF1-2   856940.0   \n",
       "1839                            uncharacterized protein     NaN  4594642.0   \n",
       "\n",
       "           Gene_Type Transcripts_accession Protein_accession  Protein_length  \\\n",
       "1522  protein-coding        NM_001184302.1       NP_010835.1          1277.0   \n",
       "1523  protein-coding        NM_001180855.1       NP_010836.1           216.0   \n",
       "1524  protein-coding        NM_001178891.1       NP_010837.1           216.0   \n",
       "1525  protein-coding        NM_001178890.1       NP_010839.1           122.0   \n",
       "1526  protein-coding        NM_001178888.1       NP_010841.1           107.0   \n",
       "...              ...                   ...               ...             ...   \n",
       "1835  protein-coding        NM_001348829.1    NP_001335771.1           239.0   \n",
       "1836  protein-coding        NM_001184633.1       NP_878070.1            99.0   \n",
       "1837  protein-coding        NM_001179079.3       NP_011116.3           122.0   \n",
       "1838  protein-coding        NM_001179080.3       NP_011117.3          1681.0   \n",
       "1839  protein-coding        NM_001184590.1    NP_001073292.1           160.0   \n",
       "\n",
       "      Locus_tag  \n",
       "1522    YEL077C  \n",
       "1523  YEL076C-A  \n",
       "1524    YEL076C  \n",
       "1525    YEL075C  \n",
       "1526    YEL073C  \n",
       "...         ...  \n",
       "1835    YER188W  \n",
       "1836  YER188C-A  \n",
       "1837    YER189W  \n",
       "1838    YER190W  \n",
       "1839  YER190C-B  \n",
       "\n",
       "[318 rows x 13 columns]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_ncbi.loc[genes_ncbi.Chromosome == \"V\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84feb160-6a98-4a66-b9ad-abbaf251bb8f",
   "metadata": {},
   "source": [
    "También se puede crear un filtro con dos condiciones: genes en el cromosoma V y cuyo producto de expresión tenga un largo de 100 o más aminoçidos:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "88d25479-d2e8-4636-87cc-a0356d699129",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Begin</th>\n",
       "      <th>End</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Name</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Gene_Type</th>\n",
       "      <th>Transcripts_accession</th>\n",
       "      <th>Protein_accession</th>\n",
       "      <th>Protein_length</th>\n",
       "      <th>Locus_tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1522</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>264</td>\n",
       "      <td>4097</td>\n",
       "      <td>V</td>\n",
       "      <td>minus</td>\n",
       "      <td>Y' element ATP-dependent helicase</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856630.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001184302.1</td>\n",
       "      <td>NP_010835.1</td>\n",
       "      <td>1277.0</td>\n",
       "      <td>YEL077C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1523</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>4185</td>\n",
       "      <td>5114</td>\n",
       "      <td>V</td>\n",
       "      <td>minus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856632.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001180855.1</td>\n",
       "      <td>NP_010836.1</td>\n",
       "      <td>216.0</td>\n",
       "      <td>YEL076C-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1524</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>4464</td>\n",
       "      <td>5114</td>\n",
       "      <td>V</td>\n",
       "      <td>minus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856633.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001178891.1</td>\n",
       "      <td>NP_010837.1</td>\n",
       "      <td>216.0</td>\n",
       "      <td>YEL076C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>5345</td>\n",
       "      <td>5713</td>\n",
       "      <td>V</td>\n",
       "      <td>minus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856634.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001178890.1</td>\n",
       "      <td>NP_010839.1</td>\n",
       "      <td>122.0</td>\n",
       "      <td>YEL075C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>7230</td>\n",
       "      <td>7553</td>\n",
       "      <td>V</td>\n",
       "      <td>minus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856636.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001178888.1</td>\n",
       "      <td>NP_010841.1</td>\n",
       "      <td>107.0</td>\n",
       "      <td>YEL073C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1834</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>566230</td>\n",
       "      <td>566655</td>\n",
       "      <td>V</td>\n",
       "      <td>plus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856937.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001179077.1</td>\n",
       "      <td>NP_011114.1</td>\n",
       "      <td>141.0</td>\n",
       "      <td>YER187W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1835</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>568040</td>\n",
       "      <td>568759</td>\n",
       "      <td>V</td>\n",
       "      <td>plus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856938.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001348829.1</td>\n",
       "      <td>NP_001335771.1</td>\n",
       "      <td>239.0</td>\n",
       "      <td>YER188W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1837</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>571155</td>\n",
       "      <td>571523</td>\n",
       "      <td>V</td>\n",
       "      <td>plus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>856939.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001179079.3</td>\n",
       "      <td>NP_011116.3</td>\n",
       "      <td>122.0</td>\n",
       "      <td>YER189W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1838</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>571480</td>\n",
       "      <td>576525</td>\n",
       "      <td>V</td>\n",
       "      <td>plus</td>\n",
       "      <td>Y' element ATP-dependent helicase protein 1 co...</td>\n",
       "      <td>YRF1-2</td>\n",
       "      <td>856940.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001179080.3</td>\n",
       "      <td>NP_011117.3</td>\n",
       "      <td>1681.0</td>\n",
       "      <td>YER190W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1839</th>\n",
       "      <td>NC_001137.3</td>\n",
       "      <td>575680</td>\n",
       "      <td>576162</td>\n",
       "      <td>V</td>\n",
       "      <td>minus</td>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4594642.0</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NM_001184590.1</td>\n",
       "      <td>NP_001073292.1</td>\n",
       "      <td>160.0</td>\n",
       "      <td>YER190C-B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>271 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Accession   Begin     End Chromosome Orientation  \\\n",
       "1522  NC_001137.3     264    4097          V       minus   \n",
       "1523  NC_001137.3    4185    5114          V       minus   \n",
       "1524  NC_001137.3    4464    5114          V       minus   \n",
       "1525  NC_001137.3    5345    5713          V       minus   \n",
       "1526  NC_001137.3    7230    7553          V       minus   \n",
       "...           ...     ...     ...        ...         ...   \n",
       "1834  NC_001137.3  566230  566655          V        plus   \n",
       "1835  NC_001137.3  568040  568759          V        plus   \n",
       "1837  NC_001137.3  571155  571523          V        plus   \n",
       "1838  NC_001137.3  571480  576525          V        plus   \n",
       "1839  NC_001137.3  575680  576162          V       minus   \n",
       "\n",
       "                                                   Name  Symbol    Gene_ID  \\\n",
       "1522                  Y' element ATP-dependent helicase     NaN   856630.0   \n",
       "1523                            uncharacterized protein     NaN   856632.0   \n",
       "1524                            uncharacterized protein     NaN   856633.0   \n",
       "1525                            uncharacterized protein     NaN   856634.0   \n",
       "1526                            uncharacterized protein     NaN   856636.0   \n",
       "...                                                 ...     ...        ...   \n",
       "1834                            uncharacterized protein     NaN   856937.0   \n",
       "1835                            uncharacterized protein     NaN   856938.0   \n",
       "1837                            uncharacterized protein     NaN   856939.0   \n",
       "1838  Y' element ATP-dependent helicase protein 1 co...  YRF1-2   856940.0   \n",
       "1839                            uncharacterized protein     NaN  4594642.0   \n",
       "\n",
       "           Gene_Type Transcripts_accession Protein_accession  Protein_length  \\\n",
       "1522  protein-coding        NM_001184302.1       NP_010835.1          1277.0   \n",
       "1523  protein-coding        NM_001180855.1       NP_010836.1           216.0   \n",
       "1524  protein-coding        NM_001178891.1       NP_010837.1           216.0   \n",
       "1525  protein-coding        NM_001178890.1       NP_010839.1           122.0   \n",
       "1526  protein-coding        NM_001178888.1       NP_010841.1           107.0   \n",
       "...              ...                   ...               ...             ...   \n",
       "1834  protein-coding        NM_001179077.1       NP_011114.1           141.0   \n",
       "1835  protein-coding        NM_001348829.1    NP_001335771.1           239.0   \n",
       "1837  protein-coding        NM_001179079.3       NP_011116.3           122.0   \n",
       "1838  protein-coding        NM_001179080.3       NP_011117.3          1681.0   \n",
       "1839  protein-coding        NM_001184590.1    NP_001073292.1           160.0   \n",
       "\n",
       "      Locus_tag  \n",
       "1522    YEL077C  \n",
       "1523  YEL076C-A  \n",
       "1524    YEL076C  \n",
       "1525    YEL075C  \n",
       "1526    YEL073C  \n",
       "...         ...  \n",
       "1834    YER187W  \n",
       "1835    YER188W  \n",
       "1837    YER189W  \n",
       "1838    YER190W  \n",
       "1839  YER190C-B  \n",
       "\n",
       "[271 rows x 13 columns]"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_ncbi.loc[(genes_ncbi.Chromosome == \"V\") & (genes_ncbi.Protein_length >= 100)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07b720c2-3378-491e-9f48-165bab7eef27",
   "metadata": {},
   "source": [
    "## Fusión de DataFrames"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba00fe68-f013-4b10-affd-4f72f271f634",
   "metadata": {},
   "source": [
    "A veces necesitamos combinar información de dos DataFrames diferentes. Por ejemplo, en nuestro caso, el DataFrame genes_crec contiene datos de experimentos de crecimiento de cepas de *S. cerevisiae* con diferentes medios de cultivo. No sería raro que una vez que tenemos esos datos experimentales nos interese enriquecer la información de esos genes. Para consegiur esto vamos a usar el otro DataFrame de este ejercicio, genes_ncbi.\n",
    "\n",
    "Primero vamos a revisar y repasar la información que tienen estos DataFrames en cuanto a identificadores de genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "985c00c1-7407-47f8-a5d0-92c8f500bb20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Accession', 'Begin', 'End', 'Chromosome', 'Orientation', 'Name',\n",
       "       'Symbol', 'Gene_ID', 'Gene_Type', 'Transcripts_accession',\n",
       "       'Protein_accession', 'Protein_length', 'Locus_tag'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_ncbi.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "0480f9d6-e204-49db-9fcd-e9d017690860",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Locus_tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>seripauperin PAU8</td>\n",
       "      <td>PAU8</td>\n",
       "      <td>YAL068C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAL067W-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>putative permease SEO1</td>\n",
       "      <td>SEO1</td>\n",
       "      <td>YAL067C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAL065C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>uncharacterized protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAL064W-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tda8p</td>\n",
       "      <td>TDA8</td>\n",
       "      <td>YAL064C-A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Name Symbol  Locus_tag\n",
       "0        seripauperin PAU8   PAU8    YAL068C\n",
       "1  uncharacterized protein    NaN  YAL067W-A\n",
       "2   putative permease SEO1   SEO1    YAL067C\n",
       "3  uncharacterized protein    NaN    YAL065C\n",
       "4  uncharacterized protein    NaN  YAL064W-B\n",
       "5                    Tda8p   TDA8  YAL064C-A"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_ncbi.loc[0:5, ['Name', 'Symbol', 'Locus_tag']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "27a1f32f-ad6f-4cc1-b0e5-ed61582782ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ORF', 'Growth_20g', 'Growth_60g', 'Growth_MM', 'Growth_YPD_NaCl',\n",
       "       'Growth_YP_lactate'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_crec.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "0b514a90-1c02-4d8b-9e77-ff5927bad066",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       YAL003W\n",
       "1       YAL004W\n",
       "2       YAL008W\n",
       "3       YAL013W\n",
       "4       YAL014C\n",
       "         ...   \n",
       "3653    YPR196W\n",
       "3654    YPR197C\n",
       "3655    YPR198W\n",
       "3656    YPR199C\n",
       "3657    YPR200C\n",
       "Name: ORF, Length: 3658, dtype: object"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_crec['ORF']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfed3953-e961-4fb2-916e-71c6b544b9a4",
   "metadata": {},
   "source": [
    "La columna Locus_tag del DataFrame genes_ncbi tiene los nombres oficiales de los loci del genoma de la levadura, y lo mismo ocurre con la columna ORF del otro DataFrame. Esto significa que podemos unir ambos DataFrames usando esa columna. Pero hay otra consideración más. La información del NCBI contiene información de muchos genes que en este momento no nos interesan, mientras que para el DataFrame genes_crec queremos conservar toda la información -porque son nuestros datos experimentales-, aún cuando el NCBI no tuviera información para alguno de los loci.\n",
    "\n",
    "Para fusionar ambos DataFrames tenemos que especificar qué columnas usar para parear registros y qué DataFrame queremos dejar con todos sus registros. En este caso, fusionamos genes_crec con genes_ncbi. Es decir, genes_crec queda a la izquierda y es el DataFrame que queremos conservar completo (la alternativa sería fusionar genes_ncbi con genes_crec):\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "595b6989-b082-42a4-bf69-a1ada1e6aaec",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_crec_info = genes_crec.merge(genes_ncbi, \n",
    "                                   how = 'left', \n",
    "                                   left_on = 'ORF', \n",
    "                                   right_on = 'Locus_tag')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "6f5f0fe4-a67f-44b0-8c9a-b11756a206da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          ORF Growth_20g Growth_60g Growth_MM Growth_YPD_NaCl  \\\n",
      "0     YAL003W        Ess        Ess       Ess             Ess   \n",
      "1     YAL004W        Non        Non        ND             Non   \n",
      "2     YAL008W        Non        Non       Non             Non   \n",
      "3     YAL013W        Non        0.6        ND             Non   \n",
      "4     YAL014C        0.8        0.2       0.3             0.4   \n",
      "...       ...        ...        ...       ...             ...   \n",
      "3653  YPR196W        0.5        0.8       0.7             Non   \n",
      "3654  YPR197C        Non        Ess       0.9             Non   \n",
      "3655  YPR198W        0.7        0.6       Non             0.7   \n",
      "3656  YPR199C        0.3        0.2       0.1             0.6   \n",
      "3657  YPR200C        0.5        Non       0.8             0.6   \n",
      "\n",
      "     Growth_YP_lactate    Accession     Begin       End Chromosome  \\\n",
      "0                  Ess  NC_001133.9  142174.0  143160.0          I   \n",
      "1                  Non          NaN       NaN       NaN        NaN   \n",
      "2                  Non  NC_001133.9  136914.0  137510.0          I   \n",
      "3                  0.8  NC_001133.9  129270.0  130487.0          I   \n",
      "4                  0.4  NC_001133.9  128252.0  129019.0          I   \n",
      "...                ...          ...       ...       ...        ...   \n",
      "3653               0.8  NC_001148.4  931376.0  932788.0        XVI   \n",
      "3654               Non          NaN       NaN       NaN        NaN   \n",
      "3655               0.8  NC_001148.4  934034.0  935665.0        XVI   \n",
      "3656               0.4  NC_001148.4  938148.0  939032.0        XVI   \n",
      "3657               0.4  NC_001148.4  939279.0  939671.0        XVI   \n",
      "\n",
      "     Orientation                                              Name Symbol  \\\n",
      "0           plus      translation elongation factor 1 subunit beta   EFB1   \n",
      "1            NaN                                               NaN    NaN   \n",
      "2           plus                                            Fun14p  FUN14   \n",
      "3           plus    Rpd3L histone deacetylase complex subunit DEP1   DEP1   \n",
      "4          minus                                          syntaxin   SYN8   \n",
      "...          ...                                               ...    ...   \n",
      "3653        plus  putative maltose-responsive transcription factor    NaN   \n",
      "3654         NaN                                               NaN    NaN   \n",
      "3655        plus                                             Sge1p   SGE1   \n",
      "3656       minus                                             Arr1p   ARR1   \n",
      "3657       minus                                             Arr2p   ARR2   \n",
      "\n",
      "       Gene_ID       Gene_Type Transcripts_accession Protein_accession  \\\n",
      "0     851260.0  protein-coding        NM_001178150.1       NP_009398.1   \n",
      "1          NaN             NaN                   NaN               NaN   \n",
      "2     851225.0  protein-coding        NM_001178153.1       NP_009394.1   \n",
      "3     851220.0  protein-coding        NM_001178158.2       NP_009389.3   \n",
      "4     851219.0  protein-coding        NM_001178159.1       NP_009388.2   \n",
      "...        ...             ...                   ...               ...   \n",
      "3653  856326.0  protein-coding        NM_001184293.1       NP_015522.1   \n",
      "3654       NaN             NaN                   NaN               NaN   \n",
      "3655  856327.0  protein-coding        NM_001184295.1       NP_015524.1   \n",
      "3656  856329.0  protein-coding        NM_001184296.1       NP_015525.1   \n",
      "3657  856330.0  protein-coding        NM_001184297.1       NP_015526.1   \n",
      "\n",
      "      Protein_length Locus_tag  \n",
      "0              206.0   YAL003W  \n",
      "1                NaN       NaN  \n",
      "2              198.0   YAL008W  \n",
      "3              405.0   YAL013W  \n",
      "4              255.0   YAL014C  \n",
      "...              ...       ...  \n",
      "3653           470.0   YPR196W  \n",
      "3654             NaN       NaN  \n",
      "3655           543.0   YPR198W  \n",
      "3656           294.0   YPR199C  \n",
      "3657           130.0   YPR200C  \n",
      "\n",
      "[3658 rows x 19 columns]\n"
     ]
    }
   ],
   "source": [
    "print(genes_crec_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "26f3b13d-cc01-49a5-975a-9bc2c3335cba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6477, 13)"
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_ncbi.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "ff1d53d0-13b0-4cef-83c5-5cff801bf192",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3658, 6)"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_crec.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "489f0f1b-c5ea-43bc-acba-c2ad484b676b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3658, 19)"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_crec_info.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1757b79d-56a0-45b8-bf3a-3cfd20a0b566",
   "metadata": {},
   "source": [
    "## Sugerencias de ejercicios"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb9ff26d-8af6-48af-9c9d-5725abd12004",
   "metadata": {},
   "source": [
    "Los dos DataFrames tienen bastante información y se podrían hacer otros filtrados. Por ejemplo, las columnas de genes_crec pueden tomar valores numéricos, indicando el efecto sobre el crecimiento con respecto a la cepa *wild type*, o el valor texto \"ess\"para indicar que es un gen esencial, o \"Non\", porque es un gen no esencial y no se observa ningún cambio en el crecimiento, o \"ND\", indicando que no se realizó la medición. De manera que las columnas son de tipo string, pero se pueden filtrar los genes esenciales, no esenciales y en tercer lugar los de efecto intermedio convirtiendolos a números. \n",
    "\n",
    "No lo vimos en esta clase, pero los DataFrames también se pueden ordenar de acuerdo a alguna columna. es algo sencillo de hacer y pueden buscar ustedes la información sobre cómo hacerlo y proponer ejemplos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0708c19a-bf4c-413a-bc36-a8062cc2486d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
